File: progress.py
import sys
import time

TOTAL_DURATION = 60  # Total duration of the progress bar in seconds
STEPS = 50  # Number of steps in the progress bar

def main():
    interval = TOTAL_DURATION / STEPS
    for step in range(STEPS + 1):
        progress = "=" * step + " " * (STEPS - step)
        percentage = step * 100 // STEPS
        remaining_time = (STEPS - step) * interval
        eta_minutes = int(remaining_time // 60)
        eta_seconds = int(remaining_time % 60)
        sys.stdout.write(f"\r[{progress}] {percentage:3}% (ETA: {eta_minutes:02}:{eta_seconds:02})")
        sys.stdout.flush()
        time.sleep(interval)

    print("\nDone!")

if __name__ == "__main__":
    main()

File: progress.rs
use std::thread;
use std::time::{Duration, Instant};
use std::io::{self, Write}; // Import Write trait

const DURATION: u64 = 60; // Total duration of the progress bar in seconds

fn main() {
    let interval = Duration::from_secs(DURATION) / 50; // Dividing total duration by number of steps (50)
    let steps = 50;

    let mut stdout = io::stdout(); // Obtain stdout

    for step in 0..=steps {
        print!("\r[");
        for _ in 0..step {
            print!("=");
        }
        for _ in step..steps {
            print!(" ");
        }
        print!("] {:3}% ", step * 100 / steps);
        let remaining_time = Duration::from_secs((steps - step) as u64 * interval.as_secs());
        print!("(ETA: {:02}:{:02})", remaining_time.as_secs() / 60, remaining_time.as_secs() % 60);
        stdout.flush().unwrap(); // Flush stdout
        thread::sleep(interval);
    }

    println!("\nDone!");
}
